botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,275 @@
+        """
+
+        super(MCAcquisitionFunction, self).__init__(model=model)
+
+        if design_sampler is None:
+            if n_fantasy_at_design_pts is None:
+                raise ValueError((
+                    "Must specify `n_fantasy_at_design_pts` if no `design_sampler` "
+                    "is provided."
+                ))
+            # base samples should be fixed for joint optimization over X, A
+            design_sampler = SobolQMCNormalSampler(
+                sample_shape=n_fantasy_at_design_pts,
+                resample=False,
+                collapse_batch_dims=True,
+            )
+        elif n_fantasy_at_design_pts is not None:
+            if design_sampler.sample_shape != torch.Size([n_fantasy_at_design_pts]):
+                raise ValueError((
+                    "The design_sampler shape must match n_fantasy_at_design_pts="
+                    f"{n_fantasy_at_design_pts}."
+                ))
+        else:
+            n_fantasy_at_design_pts = design_sampler.sample_shape[0]
+
+        if action_sampler is None:
+            if n_fantasy_at_action_pts is None:
+                raise ValueError((
+                    "Must specify `n_fantasy_at_action_pts` if no `action_sampler` "
+                    "is provided."
+                ))
+            # base samples should be fixed for joint optimization over X, A
+            action_sampler = SobolQMCNormalSampler(
+                sample_shape=n_fantasy_at_action_pts,
+                resample=False,
+                collapse_batch_dims=True,
+            )
+        elif n_fantasy_at_action_pts is not None:
+            if action_sampler.sample_shape != torch.Size([n_fantasy_at_action_pts]):
+                raise ValueError((
+                    "The sampler shape must match n_fantasy_at_action_pts="
+                    f"{n_fantasy_at_action_pts}."
+                ))
+        else:
+            n_fantasy_at_action_pts = action_sampler.sample_shape[0]
+
+        self.design_sampler = design_sampler
+        self.action_sampler = action_sampler
+        self.n_fantasy_at_design_pts = n_fantasy_at_design_pts
+        self.loss_function_hyperparameters = loss_function_hyperparameters
+        self.loss_function = loss_function_class(
+            loss_function_hyperparameters=self.loss_function_hyperparameters,
+        )
+
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor, A: Tensor) -> Tensor:
+        r"""Evaluate qHEntropySearch objective (q-HES) on the candidate set `X`.
+
+        Args:
+            X: Design tensor of shape `batch x q x num_dim_design`.
+            A: Action tensor of shape `batch x n_fantasy_at_design_pts
+                x num_actions x num_dim_action`.
+
+        Returns:
+            A Tensor of shape `batch`.
+        """
+
+        # construct the fantasy model of shape `n_fantasy_at_design_pts x b`
+        model = copy.deepcopy(self.model)
```

Comment:
Thank you for the question. It appears that the issue is related to the gradient of X being retained by self.model.fantasize even after calling optimizer.zero_grad(), leading to an error later on ("Trying to backward through the graph a second time..."). To address this issue, we have two potential solutions: either create a deep copy of self.model using model = copy.deepcopy(self.model), or set retain_graph=True when calling loss.backward(). I believe the latter approach may have a higher memory consumption. Do you happen to know of any alternative approaches that can avoid the need for both deep copy and saving the graph?

Reply:
Hi @Balandat! I hope you're having a great day. If you have a moment, could you take a look at my question posted above? I appreciate any suggestions you may have. Thank you!

Reply:
Sorry for the late response, I was out last week.

What exactly is the code you're running when you're observing this behavior? Do you have a repro?

Note that in other acquisition functions that call `fantasize` we don't need to deepcopy the model or set `retain_graph=True`, see e.g. for the Knowledge gradient here: https://github.com/pytorch/botorch/blob/main/botorch/acquisition/knowledge_gradient.py#L194-L197

Reply:
Thank you for the reply. This behavior is observed when running the [corresponding tutorial of H-Entropy Search](https://github.com/sangttruong/botorch/blob/sang-1733-adding-hentropy-search/tutorials/hentropy_search.ipynb). I'm aware that Knowledge Gradient doesn't need a deep copy or retain graph, but I wasn't sure why this was possible in KG and not in HES.

Reply:
I talked to @Ryan-Rhys today about this and he suggests that I post the error message here. I include below the output when turning off both deep copy and retain_graph when running the corresponding tutorial. It's just a generic autograd error that isn't very helpful when debugging the acquisition function. I remember talking with @willieneis about a year ago about this error, but I don't think we have a better way to address this, other than either deep copy or setting retain_graph=True during backward, as the error message suggested. 

```
BO round: 0, Optim round: 0, Loss: -1311.93
Traceback (most recent call last):
  File "hentropy_search.py", line 256, in <module>
    from botorch.acquisition.hentropy_search import (
  File "/dfs/user/sttruong/botorch/botorch/acquisition/hentropy_search.py", line 486, in <module>
    loss.backward()
  File "/dfs/user/sttruong/env/geometric_ampere/lib/python3.8/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/dfs/user/sttruong/env/geometric_ampere/lib/python3.8/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.

```

Reply:
So I guess `torch.optim` is the issue 🤔

Reply:
@Balandat @esantorella  Hi Max and Elizabeth! Have you had a chance to look further into this? It seems like we always have this issue when using SGD `torch.optim` with acquisition functions that call `fantasize`, including the KG acquisition function. The reason KG works fine on the tutorial on the BoTorch page is that it uses a built-in optimizer in BoTorch. Please let me know if we should keep either the `deepcopy` call or using `retain_graph = True` when calling the optimizer. Thank you so much!

Reply:
@sangttruong I took a look at this - I believe the reason why autograd is complaining is that we are backpropagating through the train-train covariance cache of the base (non-fantasized) model when calling `backward()`. In particular, in the very first `backward()` call this  doesn't happen due to the fact that the caches haven't been filled. The reason this doesn't cause any issues with the botorch optimizers is b/c under the hood we're calling `torch.autograd.grad(loss, X)` rather than `loss.backward()`, so we aren't backpropagating through the train-train cache of the model (since they do not depend on `X`). I am not an expert on the `torch.optim` module internals, but it seems somewhat wasteful to compute gradients for variables that aren't optimization variables in the optimizer. I wonder if there is a way to do this?

In the meantime setting `retain_graph=True` should "fix" the problem - though it seems kind of wasteful since we don't actually need to backprop twice through anything. But it seems a lot less wasteful than deepcopying the whole model. You can also just benchmark the runtime of the two versions. 

Let me know how ii goes!